### PR TITLE
enhance handling prune for CR

### PR DESCRIPTION
**Pull request by xuzhenglun**
**What this PR does / why we need it**:

1. allow CR to decide whether to enable pruning resources;
2. allow CR to decide what kinds of resources should be pruned.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

introduce 2 interfaces to let CR can decide whether and how to prune resources. When none of them are implemented, the behavior should be the same as before. So I think this is not a breaking change.

**Additional documentation**:

### CODE REVIEW

**Comment by xuzhenglun** [comment]
ping @atoato88

This PR enhances the ability of `prune`, and it doesn't relate with `DirectApplier`.  I would be very grateful if you could take a look at this PR.

**Comment by atoato88** [comment]
I agree with this PR's enhancement about `prune-whitelist`.

I think `kubebuilder-declarative-pattern` has similar functions by [`WithApplyPrune`](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/master/pkg/patterns/declarative/options.go#L108-L113) already.
I think this framework chosed implement `With***` functions for manipurating, so it's prefer to update `WithApplyPrune` or add new `With***` functions if you would like to enhance `prune` functionality.
e.g. implement new `WithPruneWhitelist`?
WDYT?

**Comment by xuzhenglun** [comment]
> agree

`With***` function works just fine when we want to build a specific addon operator.  However, I'm trying to implement a generic AddonManager for Cluster.  My user could choose to enable `pruning` by setting a flag in CR. Feature more, consisting of Addon could have CRD inside, so they should be able to configure the white list to determine what kind of resource could be pruned.

To achieve the above by using the `WithPrune` function, It's necessary to pass CR to the user's function, and it will change the function signature of `WithPrune` which can be a breaking change.  Secondly, framework users still need to implement methods like `Prune` or `PruneWhiteList` anyway, because those configurations are defined in CR.

WDYT?

File: pkg/patterns/declarative/reconciler.go [hunk 1]
```diff
@@ -74,6 +74,14 @@ type DeclarativeObject interface {
 	metav1.Object
 }
 
+type Pruner interface {
```
Comment (atoato88) [inline pkg/patterns/declarative/reconciler.go:77]: Could you add some explaination for this interface?

File: pkg/patterns/declarative/reconciler.go [hunk 2]
```diff
@@ -74,6 +74,14 @@ type DeclarativeObject interface {
...
+	Prune() bool
+}
+
+type PruneWhiteLister interface {
```
Comment (atoato88) [inline pkg/patterns/declarative/reconciler.go:81]: ditto.

**Comment by atoato88** [comment]
Thank you to explain the context detailed.
I agree with your thinking, I think it's preferable for achieve your requirements.

BTW, I left review comments above, please check those.

File: pkg/patterns/declarative/reconciler.go (hunk 2 above)
Comment (atoato88) [reply pkg/patterns/declarative/reconciler.go:81]: And, `[]string` has specific format for `gvk`?
If so, please wirte that.

File: pkg/patterns/declarative/reconciler.go (hunk 1 above)
Comment (xuzhenglun) [reply pkg/patterns/declarative/reconciler.go:80]: An explanation of the interface has been appended.

**Comment by xuzhenglun** [comment]
> Thank you to explain the context detailed. I agree with your thinking, I think it's preferable for achieve your requirements.
> 
> BTW, I left review comments above, please check those.

Thanks for your review. I append some comments in the source code and just updated this PR.

File: pkg/patterns/declarative/reconciler.go (hunk 2 above)
Comment (atoato88) [reply pkg/patterns/declarative/reconciler.go:88]: perfect! :smile:

**Comment by atoato88** [comment]
/approve
Thank you to update quickly!

**Comment by atoato88** [comment]
/approve cancel
oh, we should check test before approving.

**Comment by xuzhenglun** [comment]
@atoato88 still needs an LGTM label to merge. Is something I need to do yet?

**Pull request merged**